first-party/jib-quarkus-extension-gradle/src/main/java/com/google/cloud/tools/jib/gradle/extension/quarkus/JibQuarkusExtension.java

```diff
@@ -0,0 +1,178 @@
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.gradle.api.Project;
+import org.gradle.api.artifacts.ResolvedArtifact;
+import org.gradle.api.artifacts.component.ProjectComponentIdentifier;
+import org.gradle.api.plugins.JavaPlugin;
+import org.gradle.jvm.tasks.Jar;
+
+public class JibQuarkusExtension implements JibGradlePluginExtension<Void> {
+
+  private AbsoluteUnixPath appRoot = AbsoluteUnixPath.get("/app");
+  private List<String> jvmFlags = Collections.emptyList();
+
+  @Override
+  public Optional<Class<Void>> getExtraConfigType() {
+    return Optional.empty();
+  }
+
+  @Override
+  public ContainerBuildPlan extendContainerBuildPlan(
+      ContainerBuildPlan buildPlan,
+      Map<String, String> properties,
+      Optional<Void> config,
+      GradleData gradleData,
+      ExtensionLogger logger)
+      throws JibPluginExtensionException {
+    try {
+      logger.log(LogLevel.LIFECYCLE, "Running Quarkus Jib extension");
+      readJibConfigurations(gradleData.getProject());
+
+      Project project = gradleData.getProject();
+      Path buildDir = project.getBuildDir().toPath();
+      Jar jarTask = (Jar) project.getTasks().findByName("jar");
+      String jarName = jarTask.getArchiveFile().get().getAsFile().getName();
+      Path jar = buildDir.resolve(jarName.replaceAll("\\.jar$", "-runner.jar"));
+
+      if (!Files.isRegularFile(jar)) {
+        throw new JibPluginExtensionException(
+            getClass(),
+            jar + " doesn't exist; did you run the Qaurkus Gradle plugin ('quarkusBuild' task)?");
+      }
+
+      ContainerBuildPlan.Builder planBuilder = buildPlan.toBuilder();
+      planBuilder.setLayers(Collections.emptyList());
+
+      // dependency layers
+      addDependencyLayers(project, planBuilder, buildDir.resolve("lib"));
+
+      // Quarkus runner JAR layer
+      AbsoluteUnixPath appRootJar = appRoot.resolve("app.jar");
+      FileEntriesLayer jarLayer =
+          FileEntriesLayer.builder().setName("quarkus jar").addEntry(jar, appRootJar).build();
+      planBuilder.addLayer(jarLayer);
+
+      // Preserve extra directories layers.
+      buildPlan
+          .getLayers()
+          .stream()
+          .filter(layer -> layer.getName().startsWith("extra files"))
+          .forEach(planBuilder::addLayer);
+
+      // set entrypoint
+      List<String> entrypoint = new ArrayList<>();
+      entrypoint.add("java");
+      entrypoint.addAll(jvmFlags);
+      entrypoint.add("-jar");
+      entrypoint.add(appRootJar.toString());
+      planBuilder.setEntrypoint(entrypoint);
+
+      return planBuilder.build();
+
+    } catch (IOException ex) {
+      throw new JibPluginExtensionException(getClass(), Verify.verifyNotNull(ex.getMessage()), ex);
+    }
+  }
+
+  @VisibleForTesting
+  void addDependencyLayers(
+      Project project, ContainerBuildPlan.Builder planBuilder, Path libDirectory)
```

Comment:
I thought quarkus put all dependency information in the MANIFEST? Could we not just get that information from there? They should have a dependency copy task somewhere? I remember having slightly different behaviors for some dependencies (or like they were included by the plugin - maybe some quarkus runtime stuff?)

Reply:
> They should have a dependency copy task somewhere?

They do this and these are placed into `<build_out>/lib` (after augmenting and renaming). And yes, since the standard Jib build retrieves the dep jars from Maven/Gradle, it didn't work. However, I believe this extension should work fine. As discussed, we can go with this first implementation and eventually move to your idea to get the info from MANIFEST. At that point, we should be able to create another general extension to do `java -jar`.

Reply:
Actually, I just looked into the default `Dockerfile.jvm` that Quarkus app generator generated, and it just copies the runner jar and all the jars from `lib/`, which is exactly what this extension does.

```Dockerfile
COPY target/lib/* /deployments/lib/
COPY target/*-runner.jar /deployments/app.jar
```